### ISSUE TITLE
[1.4.1] Cuda build fails

### ISSUE THREAD

**Issue by tuxmaster5000** [pytorch/pytorch#36721]
Using:
Cuda 10.2
GCC:  7.3.1
Without cuda  the build will be done.
```
[ 23%] Building CXX object third_party/gloo/gloo/CMakeFiles/gloo.dir/common/logging.cc.o
c++: error: unrecognized command line option '-faligned-new'

```
[build.log.zip](https://github.com/pytorch/pytorch/files/4486925/build.log.zip)

**Comment by mruberry** [comment]
Can you provide additional details about your environment and the commands you ran to produce this issue? Looking at the build log, it's not clear where this error is coming from.

**Comment by tuxmaster5000** [comment]
Yes, it will happens on every build. Because the build is done in an chroot environment, I can reproduce it every time. What type of information do you need?

**Comment by mruberry** [comment]
It'd be helpful to know, for example, the operating system and which instructions to build you were following.

Let me put it another way: can you build the PyTorch master branch from source following the github instructions?

**Comment by tuxmaster5000** [comment]
I use CentOS 7.7 and the mock tool, to build rpm packages to distribute it on the servers.
I can try the master branch, but the instruction will not work for CentOS because conda is not available here. So I can only check out the master, call git submodule update and try to build it via python3 setup.py install in the chroot environment. Like it was shown in the log.

**Comment by malfet** [comment]
@tuxmaster5000, does you host compiler supports `-faligned-new`? I.e. can you run `printf '#include <iostream>\nint main() { std::cout << new int[5] << std::endl; }'|c++ -x c++ -faligned-new -; ./a.out` and share the output here?

P.S. Compilation flag is added here for all `gcc` version greater than 7.0. https://github.com/pytorch/pytorch/blob/v1.4/CMakeLists.txt#L458

**Comment by tuxmaster5000** [comment]
gcc -v:

> Using built-in specs.
> COLLECT_GCC=gcc
> COLLECT_LTO_WRAPPER=/opt/rh/devtoolset-7/root/usr/libexec/gcc/x86_64-redhat-linux/7/lto-wrapper
> Target: x86_64-redhat-linux
> Configured with: ../configure --enable-bootstrap --enable-languages=c,c++,fortran,lto --prefix=/opt/rh/devtoolset-7/root/usr --mandir=/opt/rh/devtoolset-7/root/usr/share/man --infodir=/opt/rh/devtoolset-7/root/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-shared --enable-threads=posix --enable-checking=release --enable-multilib --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-gcc-major-version-only --enable-plugin --with-linker-hash-style=gnu --enable-initfini-array --with-default-libstdcxx-abi=gcc4-compatible --with-isl=/builddir/build/BUILD/gcc-7.3.1-20180303/obj-x86_64-redhat-linux/isl-install --enable-libmpx --enable-gnu-indirect-function --with-tune=generic --with-arch_32=i686 --build=x86_64-redhat-linux
> Thread model: posix
> gcc version 7.3.1 20180303 (Red Hat 7.3.1-5) (GCC) 

printf '#include <iostream>\nint main() { std::cout << new int[5] << std::endl; }'|LANG=C gcc -x c++ ```
-faligned-new -; ./a.out:
/tmp/cccBkFYO.o: In function `main':
:(.text+0xa): undefined reference to `operator new[](unsigned long)'
:(.text+0x12): undefined reference to `std::cout'
:(.text+0x17): undefined reference to `std::ostream::operator<<(void const*)'
:(.text+0x1c): undefined reference to `std::basic_ostream<char, std::char_traits<char> >& std::endl<char, std::char_traits<char> >(std::basic_ostream<char, std::char_traits<char> >&)'
:(.text+0x24): undefined reference to `std::ostream::operator<<(std::ostream& (*)(std::ostream&))'
/tmp/cccBkFYO.o: In function `__static_initialization_and_destruction_0(int, int)':
:(.text+0x52): undefined reference to `std::ios_base::Init::Init()'
:(.text+0x61): undefined reference to `std::ios_base::Init::~Init()'
collect2: error: ld returned 1 exit status
```
I hope it will help.

**Comment by malfet** [comment]
@tuxmaster5000 in the logs you've shared `c++`(namely `/usr/lib64/ccache/c++`) rather than `gcc` is used as host C++ compiler, with the path set to `/usr/lib64/ccache:/opt/rh/devtoolset-7/root/usr/bin:/usr/lib64/ccache:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/sbin:/builddir/.local/bin:/builddir/bin`. Can you please check if `-falign-new` is indeed supported in that configuration.

And would commenting out this option in `CMakeLists.txt` solve the build problems?

**Comment by tuxmaster5000** [comment]
/usr/lib64/ccache/c++ is only an symlink, so that the command listed as c++ is very strange.
Some lines before the exits another error:
Scanning dependencies of target gloo
cc1plus: error: -Werror=dangling-else: no option -Wdangling-else

Exists an option to show which command cmake realy call with the full path?

**Comment by malfet** [comment]
@tuxmaster5000, yes you can run `cmake --trace-expand`, but what you want to do here is see the commands `make` is executing, which can be achieved by invoking `make VERBOSE=1 torch_cpu`

**Comment by tuxmaster5000** [comment]
Ok, but how to pass this to the python command?
Because python calls cmake and make.
